pkg/ovsdb/ovsdb.go

```diff
@@ -0,0 +1,450 @@
+	}
+
+	// Setup state
+	ovsDriver.ovsClient = ovs
+	ovsDriver.OvsBridgeName = bridgeName
+	ovsDriver.ovsdbCache = make(map[string]map[string]libovsdb.Row)
+
+	go func() {
+		// Register for notifications
+		ovs.Register(ovsDriver)
+
+		// Populate initial state into cache
+		initial, _ := ovs.MonitorAll("Open_vSwitch", "")
+		ovsDriver.populateCache(*initial)
+	}()
+
+	// HACK: sleep the main thread so that Cache can be populated
+	// Add a retry loop here and not wait 1 second
+	time.Sleep(1 * time.Second)
+
+	if !ovsDriver.IsBridgePresent(bridgeName) {
+		return nil, fmt.Errorf("failed to find bridge %s", bridgeName)
+	}
+
+	// Return the new OVS driver
+	return ovsDriver, nil
+}
+
+// Populate local cache of ovs state
+func (self *OvsDriver) populateCache(updates libovsdb.TableUpdates) {
+	// lock the cache for write
+	self.lock.Lock()
+	defer self.lock.Unlock()
+
+	for table, tableUpdate := range updates.Updates {
+		if _, ok := self.ovsdbCache[table]; !ok {
+			self.ovsdbCache[table] = make(map[string]libovsdb.Row)
+
+		}
+		for uuid, row := range tableUpdate.Rows {
+			empty := libovsdb.Row{}
+			if !reflect.DeepEqual(row.New, empty) {
+				self.ovsdbCache[table][uuid] = row.New
+			} else {
+				delete(self.ovsdbCache[table], uuid)
+			}
+		}
+	}
+}
+
+// Dump the contents of the cache into stdout
+func (self *OvsDriver) PrintCache() {
+	// lock the cache for read
+	self.lock.RLock()
+	defer self.lock.RUnlock()
+
+	fmt.Printf("OvsDB Cache: \n")
+
+	// walk the local cache
+	for tName, table := range self.ovsdbCache {
+		fmt.Printf("Table: %s\n", tName)
+		for uuid, row := range table {
+			fmt.Printf("  Row: UUID: %s\n", uuid)
+			for fieldName, value := range row.Fields {
+				fmt.Printf("    Field: %s, Value: %+v\n", fieldName, value)
+			}
+		}
+	}
+}
+
+// Get the UUID for root
+func (self *OvsDriver) getRootUuid() libovsdb.UUID {
+	// lock the cache for read
+	self.lock.RLock()
+	defer self.lock.RUnlock()
+
+	// find the matching uuid
+	for uuid := range self.ovsdbCache["Open_vSwitch"] {
+		return libovsdb.UUID{GoUUID: uuid}
+	}
+	return libovsdb.UUID{}
+}
+
+// Wrapper for ovsDB transaction
+func (self *OvsDriver) ovsdbTransact(ops []libovsdb.Operation) error {
+	// Perform OVSDB transaction
+	reply, _ := self.ovsClient.Transact("Open_vSwitch", ops...)
+
+	if len(reply) < len(ops) {
+		return errors.New("OVS transaction failed. Unexpected number of replies")
+	}
+
+	// Parse reply and look for errors
+	for i, o := range reply {
+		if o.Error != "" && i < len(ops) {
+			return errors.New("OVS Transaction failed err " + o.Error + "Details: " + o.Details)
+		} else if o.Error != "" {
+			return errors.New("OVS Transaction failed err " + o.Error + "Details: " + o.Details)
+		}
+	}
+
+	// Return success
+	return nil
+}
+
+// **************** OVS driver API ********************
+// Not in use
+func (self *OvsDriver) CreateBridge(bridgeName string, failMode string, stp bool) error {
```

Comment:
We should never touch bridges, we just merely attach ports to existing ones. By that we make sure that the bridge is always created and configured by administrator as part of host networking setup.

Reply:
remove this function